// @applitools/eyes-universal/dist/universal-server-new.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeServer = void 0;
const os_1 = __importDefault(require("os"));
const path_1 = __importDefault(require("path"));
const core_1 = require("@applitools/core");
const logger_1 = require("@applitools/logger");
const handler_1 = require("./handler");
const socket_1 = require("./socket");
const refer_1 = require("./refer");
const tracker_1 = require("./debug/tracker");
const custom_1 = require("./spec-driver/custom");
const webdriverSpec = __importStar(require("./spec-driver/webdriver"));
const LOG_DIRNAME = (_a = process.env.APPLITOOLS_LOG_DIR) !== null && _a !== void 0 ? _a : path_1.default.resolve(os_1.default.tmpdir(), `applitools-logs`);
async function makeServer({ debug = false, shutdownMode = 'lazy', idleTimeout = 900000, // 15min
printStdout = false, ...handlerOptions } = {}) {
    var _a;
    const { server, port } = await (0, handler_1.makeHandler)({ ...handlerOptions, debug });
    console.log(port); // NOTE: this is a part of the generic protocol
    (_a = process.send) === null || _a === void 0 ? void 0 : _a.call(process, { name: 'port', payload: { port } }); // NOTE: this is a part of the js specific protocol
    if (!server) {
        console.log(`You are trying to spawn a duplicated server, use the server on port ${port} instead`);
        return;
    }
    if (!printStdout)
        process.stdout.write = () => true; // NOTE: prevent any write to stdout
    const baseLogger = (0, logger_1.makeLogger)({
        handler: { type: 'rolling file', name: 'eyes', dirname: LOG_DIRNAME },
        label: 'eyes',
        level: 'info',
        colors: false,
    });
    console.log(`Logs saved in: ${LOG_DIRNAME}`);
    baseLogger.log('Server is started');
    let idle;
    let serverClosed = false;
    if (shutdownMode === 'stdin') {
        process.stdin.resume();
        process.stdin.on('end', () => {
            server.close();
        });
    }
    else if (shutdownMode === 'lazy') {
        if (idleTimeout) {
            idle = setTimeout(() => server.close(), idleTimeout);
        }
    }
    server.on('close', () => {
        clearTimeout(idle);
        serverClosed = true;
    });
    server.on('connection', client => {
        const refer = (0, refer_1.makeRefer)();
        const socket = (0, tracker_1.withTracker)({
            debug,
            socket: (0, socket_1.makeSocket)(client, { logger: baseLogger }),
        });
        if (shutdownMode === 'lazy' && idleTimeout) {
            clearTimeout(idle);
            socket.on('close', () => {
                if (server.clients.size > 0 || serverClosed)
                    return;
                idle = setTimeout(() => server.close(), idleTimeout);
            });
        }
        const logger = baseLogger.extend({
            console: {
                log: (message) => socket.emit('Server.log', { level: 'info', message }),
                warn: (message) => socket.emit('Server.log', { level: 'warn', message }),
                error: (message) => socket.emit('Server.log', { level: 'error', message }),
                fatal: (message) => socket.emit('Server.log', { level: 'fatal', message }),
            },
        });
        logger.console.log(`Logs saved in: ${LOG_DIRNAME}`);
        socket.command('Server.getInfo', async () => {
            return { logsDir: LOG_DIRNAME };
        });
        const sdkPromise = socket.create('Core.makeCore', ({ name, version, cwd, commands, protocol }) => {
            return (0, core_1.makeCore)({
                spec: protocol === 'webdriver' ? webdriverSpec : (0, custom_1.makeSpec)({ socket, commands }),
                agentId: `eyes-universal/${name}/${require('../package.json').version}/${version}`,
                cwd,
            });
        });
        socket.command('Core.makeManager', async (options) => {
            const sdk = await sdkPromise;
            const managerRef = refer.ref(await sdk.makeManager(options));
            return managerRef;
        });
        socket.command('Core.locate', async ({ target, settings, config }) => {
            const sdk = await sdkPromise;
            return sdk.locate({ target, settings, config, logger });
        });
        socket.command('Core.getViewportSize', async ({ target }) => {
            const sdk = await sdkPromise;
            return sdk.getViewportSize({ target, logger });
        });
        socket.command('Core.setViewportSize', async ({ target, size }) => {
            const sdk = await sdkPromise;
            return sdk.setViewportSize({ target, size, logger });
        });
        socket.command('Core.closeBatch', async ({ settings }) => {
            const sdk = await sdkPromise;
            return sdk.closeBatch({ settings, logger });
        });
        socket.command('Core.deleteTest', async ({ settings }) => {
            const sdk = await sdkPromise;
            return sdk.deleteTest({ settings, logger });
        });
        socket.command('EyesManager.openEyes', async ({ manager, target, settings, config }) => {
            var _a;
            const eyes = await ((_a = refer.deref(manager)) === null || _a === void 0 ? void 0 : _a.openEyes({ target, settings, config, logger }));
            const eyesRef = refer.ref(eyes, manager);
            return eyesRef;
        });
        socket.command('EyesManager.closeManager', async ({ manager, settings }) => {
            var _a;
            return (_a = refer.deref(manager)) === null || _a === void 0 ? void 0 : _a.closeManager({ settings, logger });
        });
        socket.command('Eyes.check', async ({ eyes, target, settings, config }) => {
            var _a;
            return (_a = refer.deref(eyes)) === null || _a === void 0 ? void 0 : _a.check({ target, settings, config, logger });
        });
        socket.command('Eyes.locateText', async ({ eyes, target, settings, config }) => {
            var _a;
            return (_a = refer.deref(eyes)) === null || _a === void 0 ? void 0 : _a.locateText({ target, settings, config, logger });
        });
        socket.command('Eyes.extractText', async ({ eyes, target, settings, config }) => {
            var _a;
            return (_a = refer.deref(eyes)) === null || _a === void 0 ? void 0 : _a.extractText({ target, settings, config, logger });
        });
        socket.command('Eyes.close', async ({ eyes, settings, config }) => {
            var _a;
            const results = await ((_a = refer.deref(eyes)) === null || _a === void 0 ? void 0 : _a.close({ settings, config, logger }));
            refer.destroy(eyes);
            return results;
        });
        socket.command('Eyes.abort', async ({ eyes }) => {
            var _a;
            const results = (_a = refer.deref(eyes)) === null || _a === void 0 ? void 0 : _a.abort();
            refer.destroy(eyes);
            return results;
        });
        socket.command('Debug.checkSpecDriver', async ({ driver, commands }) => {
            return (0, core_1.checkSpecDriver)({ spec: (0, custom_1.makeSpec)({ socket, commands }), driver });
        });
        socket.command('Debug.getHistory', async () => {
            return socket.getHistory();
        });
    });
    return { port, close: () => server.close() };
}
exports.makeServer = makeServer;
