// @applitools/core/dist/utils/wait-for-lazy-load.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.waitForLazyLoad = void 0;
const snippets_1 = require("@applitools/snippets");
const execute_poll_script_1 = require("./execute-poll-script");
async function waitForLazyLoad({ driver, settings, logger, }) {
    var _a, _b, _c, _d, _e;
    logger.log('lazy loading the page before capturing a screenshot');
    const arg = {
        scrollLength: (_a = settings.scrollLength) !== null && _a !== void 0 ? _a : 300,
        waitingTime: (_b = settings.waitingTime) !== null && _b !== void 0 ? _b : 2000,
        maxAmountToScroll: (_c = settings.maxAmountToScroll) !== null && _c !== void 0 ? _c : 15000,
    };
    const scripts = {
        main: { script: snippets_1.lazyLoad, args: [[arg]] },
        poll: { script: snippets_1.lazyLoad, args: [[]] },
    };
    await (0, execute_poll_script_1.executePollScript)({
        context: driver.currentContext,
        scripts,
        settings: { executionTimeout: 5 * 60 * 1000, pollTimeout: (_e = (_d = settings.pollTimeout) !== null && _d !== void 0 ? _d : settings.waitingTime) !== null && _e !== void 0 ? _e : 2000 },
        logger,
    });
}
exports.waitForLazyLoad = waitForLazyLoad;
